FastLin Bug: AssertionError: radius must not be negative

Running a the following code results in an assertion error for FastLin that the radius must not be negative. The input set at least doesn't have a negative radius.

```
network = NeuralVerification.read_nnet("./examples/networks/issue_net.nnet")
input_set = LazySets.Hyperrectangle([0.24213346636908017, 0.1426718524591044, -0.05412884560800202], [0.5429475555253416, 0.03619432035041159, 0.9672445409150574])
output_set = LazySets.HalfSpace([0.1815863597130689, 0.36136783684495044, -0.23625640505759504, -0.09291634108723623, 0.06643377425441743], -0.16053617125039454)
problem = NeuralVerification.Problem(network, input_set, output_set)

# Solve on FastLin
println(NeuralVerification.solve(NeuralVerification.FastLin(), problem))

```

Network:
```
//Neural Network File Format by Kyle Julian, Stanford 2016.
4, 5, 5, 12,
5, 8, 12, 10, 5,
This line extraneous
-6.55e4,-6.55e4,-6.55e4,-6.55e4,-6.55e4,
6.55e4,6.55e4,6.55e4,6.55e4,6.55e4,
0.0,0.0,0.0,0.0,0.0,0.0,
1.0,1.0,1.0,1.0,1.0,1.0,
0.49431101840087344, 0.35696116041247405, -0.02020665088910123,
0.6834566101978234, -0.3818222346620397, -0.05827795215582032,
-0.9356568799489926, 0.4997908661306476, 0.25465332921433914,
0.6175532856576793, 0.6715864495255626, 0.9701322561420445,
0.1432331748732718, -0.2859909374640095, -0.13671866148780998,
-0.9138609647885372, 0.4455540281240058, -0.6805290379794542,
0.9911457845761875, -0.9027579696484676, 0.009439499624181558,
-0.6771191648366983, -0.37123529657326815, -0.5609266989751478,
-0.7903438807480256,
-0.3922341529265032,
-0.40568562485706794,
0.30878630469748014,
0.5910729063722435,
0.21997538223656532,
0.27660736527698493,
0.4925408265344471,
-0.6246090986925878, -0.37961575681613136, -0.6370369155156976, -0.6903845123370638, -0.6504325631917487, 0.5228636324414793, 0.21732295831057113, 0.8563434012118858,
0.6762613678192655, 0.28251038383317306, -0.46559693892902265, 0.16707687722990183, -0.48207230348023966, -0.5316124697297355, 0.5402057501595583, -0.16376060055175667,
0.4807176452203552, -0.2527704277656393, 0.9935969024504248, 0.29889117348244243, 0.7871242423026028, -0.6693748501341799, 0.5515402051535667, -0.5895531216456056,
-0.9729402724799519, -0.6182372137326242, -0.2810219357850805, 0.2771446698396307, -0.8277689179012144, 0.44246124760028493, -0.955356731366479, -0.6368529729735708,
-0.4172233649795811, -0.6717003869120002, -0.9706559280798492, -0.5667280711385945, -0.8101379192093274, 0.7046907121544397, 0.15384245285608555, -0.8117089347757243,
0.1275274004151492, 0.25003591949015647, -0.16179650487188546, 0.012977651488937703, -0.6105528300547856, 0.9977325882328434, 0.8579956488908649, -0.08916896667476282,
0.37019065958455943, -0.7294457317299159, 0.4025510015389493, 0.8801992329478705, 0.5301107436726049, 0.40825920646732783, 0.008561407997595971, 0.9222769298573223,
0.6614086045047833, -0.2977204470126056, -0.5913204488416013, 0.4880994992701799, 0.14367141732666466, -0.2792302239802966, -0.254904355867553, -0.1198888530817297,
-0.5656410601632578, -0.0877103106791326, -0.9317260577321096, 0.9260763451053768, -0.27578964385819926, -0.6875501414606919, -0.508612788587572, -0.8734260602458588,
-0.8006351143013006, 0.3468357113335223, 0.8333875071632986, 0.7832325237775839, -0.16204965022478834, 0.17383132881614127, 0.029122119356080667, 0.44147040255731307,
-0.7463199495518271, -0.8839346291885599, 0.3510389801666389, -0.4798341466545981, -0.000605701419607918, -0.08153171045849827, 0.9888563905853376, -0.38345108179054144,
-0.5278782209820485, -0.8942656817844727, -0.2747224311112464, -0.8811619428370494, 0.6267843903932993, -0.08916351686651058, 0.9910888741628243, -0.731086971679018,
-0.6126608160237925,
-0.032701378070306486,
-0.7253226964557342,
0.49201992725115407,
-0.11999946232895864,
0.740725395871356,
0.6741377694363804,
0.99707637010843,
0.3404608368598292,
-0.3639888524049266,
0.8655696223336649,
0.28634195977921273,
0.6323383288185394, 0.12268793279446877, 0.8449505969930828, -0.5010937385391614, -0.8856553552980735, 0.9357624925206656, 0.8666679700008548, 0.8135838139036968, 0.5921261554795372, 0.17531299826057545, 0.5732260022503732, 0.21720502736072023,
-0.6619276789136319, 0.15999650951161648, 0.8969029390514551, -0.14361194362063623, 0.5855194362073153, 0.6911725512858711, 0.982607696061403, -0.14525587422589892, 0.5769967567476084, 0.4853316141550543, -0.6278936727901558, -0.4709844251654869,
0.9440735807666676, -0.2682996895082703, -0.6554527699934156, 0.918016844292568, -0.2776020171942819, 0.6082190477494134, -0.3875020402924019, -0.937081743853398, -0.02771546237462008, -0.2450244603813312, -0.22647296208544132, 0.23329690792349655,
0.270947225529067, 0.6808112352913116, 0.2515847623371368, -0.6022494526422788, -0.38489149136200673, 0.381495228802911, -0.8458389643320179, 0.08457778058191412, -0.11322629264917294, 0.8850443694566885, -0.5256425304099195, 0.631315550037991,
0.5194554528690309, 0.9642789700014012, 0.5377633702764832, 0.6159032531365436, -0.8229988499728238, 0.5038169732731714, 0.7793157022855, -0.8180591459432094, 0.9530156363213407, -0.15254172709460168, -0.8132155780572843, 0.09135199916132475,
0.1933904337019947, 0.24883010587484122, -0.6666754611938943, 0.30682590102952023, 0.4336603533893846, 0.33927326181709594, -0.9877037046021049, -0.9198282078904318, 0.35785343770308575, 0.04309474483080189, -0.8713764253755398, 0.8267247324709461,
-0.9001991134526097, 0.28135610204360484, 0.5671107826697073, 0.8834181180386724, 0.5538144311163156, 0.7108860863167856, -0.4055023098331221, -0.2976348251575436, -0.9729923994928193, -0.21056265158378462, -0.0684656291343404, -0.9490734418822817,
-0.9084610754782685, 0.7194576170875093, -0.2819830914297117, -0.5086566448602716, 0.15987568610894654, -0.2856557802005102, 0.6565113957315964, -0.6710211590565232, -0.05402807941475585, 0.34588119222954905, 0.9656680876472712, 0.8052784118912601,
-0.22356472165577923, -0.1645979853929167, 0.11120881988523346, -0.7151959891161677, -0.2635432311811119, 0.7317816328271678, -0.15332278440446734, 0.6660320991974804, 0.7053244163850994, -0.3211252248288581, -0.49152515669722296, -0.3178181539980316,
0.34623819605029027, -0.7400704931023623, 0.9876252080318695, -0.08172322436020263, -0.8910633530965266, 0.17514051585249213, 0.8725867361607604, 0.8226643918460961, 0.8891297853357698, 0.6443168073109384, 0.6843445756367856, -0.36502525514926143,
-0.8611476093175918,
-0.24305504308782888,
0.08304715868817514,
0.859614713232475,
0.746472147475342,
-0.825395076125846,
-0.693081670545237,
-0.7993901316202967,
0.5818195921660871,
-0.7171863535281453,
0.41894213029013816, 0.43017352569028633, 0.7718963651637099, -0.07643593583520314, 0.8612530729255128, -0.12579143415161553, -0.08875335865854916, -0.30796283837095917, -0.9052063651322109, 0.9786913629697036,
-0.3009396091656722, 0.5315004229052187, 0.7777424262604535, 0.4384979922222212, -0.9857194363192638, 0.9128245597743425, -0.26075506302213425, 0.3931733365404111, 0.5334760738874524, -0.10512173663252566,
0.15567934214720847, -0.014313342058568601, -0.6667541840938331, -0.5985933201544094, 0.11398610550631805, -0.45268310265914735, -0.5272732034376157, 0.23239213356605726, -0.7286909429394863, 0.17384416528342816,
0.516170266055739, 0.8165678223815487, -0.19526249547860441, 0.5681949909394026, 0.6435054473945505, -0.1316109744128311, -0.996777163124515, -0.8017630945991137, 0.48472247065670526, 0.6834059500249974,
0.9807443084950829, 0.19114393960023301, -0.6313169153510434, 0.4017171423050918, -0.5309799087169162, 0.497287477797109, -0.5967267899775077, -0.6056292251453566, 0.7471338815213349, 0.9184870606351176,
-0.15515157805657953,
-0.9779135966512196,
-0.8332839178624272,
0.9633863712988546,
-0.8810439688867673,
```

Leads to the stacktrace: 
```
LoadError: AssertionError: radius must not be negative
Stacktrace:
 [1] macro expansion at /Users/castrong/.julia/dev/LazySets/src/Assertions/Assertions.jl:23 [inlined]
 [2] (::getfield(LazySets, Symbol("##Hyperrectangle#134#136")))(::Bool, ::Type{Hyperrectangle}, ::Array{Float64,1}, ::Array{Float64,1}) at /Users/castrong/.julia/dev/LazySets/src/Sets/Hyperrectangle.jl:91
 [3] Type at ./simdloop.jl:0 [inlined]
 [4] #Hyperrectangle#138(::Array{Float64,1}, ::Array{Float64,1}, ::Bool, ::Type{Hyperrectangle}) at /Users/castrong/.julia/dev/LazySets/src/Sets/Hyperrectangle.jl:107
 [5] Type at ./array.jl:0 [inlined]
 [6] forward_network at /Users/castrong/Desktop/Research/NV_Fork/NeuralVerification.jl/src/adversarial/fastLin.jl:63 [inlined]
 [7] solve(::NeuralVerification.FastLin, ::NeuralVerification.Problem{Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}},LazySets.HalfSpace{Float64,Array{Float64,1}}}) at /Users/castrong/Desktop/Research/NV_Fork/NeuralVerification.jl/src/adversarial/fastLin.jl:43
 [8] top-level scope at /Users/castrong/Desktop/Research/NV_Fork/NeuralVerification.jl/examples/SimpleProblem.jl:25
```

--- a/src/optimization/convDual.jl
+++ b/src/optimization/convDual.jl
@@ -76,22 +76,21 @@ end
 # This step is similar to reachability method
 function get_bounds(nnet::Network, input::Vector{Float64}, ϵ::Float64)
     layers  = nnet.layers
-    n_layer = length(layers)
 
-    l = Vector{Vector{Float64}}()
-    u = Vector{Vector{Float64}}()
-    γ = Vector{Vector{Float64}}()
-    μ = Vector{Vector{Vector{Float64}}}()
+    l = Vector{Vector{Float64}}() # Lower bound
+    u = Vector{Vector{Float64}}() # Upper bound
+    b = Vector{Vector{Float64}}() # bias
+    μ = Vector{Vector{Vector{Float64}}}() # Dual variables
     input_ReLU = Vector{Vector{Float64}}()
 
     v1 = layers[1].weights'
-    push!(γ, layers[1].bias)
+    push!(b, layers[1].bias)
     # Bounds for the first layer
     l1, u1 = input_layer_bounds(layers[1], input, ϵ)
     push!(l, l1)
     push!(u, u1)
 
-    for i in 2:n_layer
+    for i in 2:length(layers)
         n_input  = length(layers[i-1].bias)
         n_output = length(layers[i].bias)
 
@@ -99,25 +98,23 @@ function get_bounds(nnet::Network, input::Vector{Float64}, ϵ::Float64)
         push!(input_ReLU, last_input_ReLU)
         D = Diagonal(last_input_ReLU)   # a matrix whose diagonal values are the relaxed_ReLU values (maybe should be sparse?)
 
-        # Propagate existing terms
+        # Propagate existing terms by right multiplication of D*W' or left multiplication of W*D
         WD = layers[i].weights*D
-        v1 = v1 * WD' # TODO CHECK
-        map!(g -> WD*g,   γ, γ)
-
-        # Updating ν_j for all previous layers
-        for M in μ
-            map!(m -> WD*m,   M, M)
+        v1 = v1 * WD' # propagate V_1^{i-1} to V_1^{i}
+        map!(g -> WD*g,   b, b) # propagate bias        
+        for V in μ
+            map!(m -> WD*m,   V, V) # Updating ν_j for all previous layers
         end
 
         # New terms
-        push!(γ, layers[i].bias)
+        push!(b, layers[i].bias)
         push!(μ, new_μ(n_input, n_output, last_input_ReLU, WD))
 
         # Compute bounds
-        ψ = v1' * input + sum(γ)
+        ψ = v1' * input + sum(b)
         eps_v1_sum = ϵ * vec(sum(abs, v1, dims = 1))
-        neg, pos = all_neg_pos_sums(input_ReLU, l, μ, n_output)
-        push!(l,  ψ - eps_v1_sum + neg )
+        neg, pos = residual(input_ReLU, l, μ, n_output)
+        push!(l,  ψ - eps_v1_sum - neg )
         push!(u,  ψ + eps_v1_sum - pos )
     end
 
@@ -125,17 +122,17 @@ function get_bounds(nnet::Network, input::Vector{Float64}, ϵ::Float64)
 end
 
 # TODO rename function and inputs
-function all_neg_pos_sums(slopes, l, μ, n_output)
+function residual(slopes, l, μ, n_output)
     # n_output = length(last(l))
     neg = zeros(n_output)
     pos = zeros(n_output)
     # Need to debug
     for (i, ℓ) in enumerate(l)                # ℓ::Vector{Float64}
-        for (j, M) in enumerate(μ[i])         # M::Vector{Float64}
+        for (j, V) in enumerate(μ[i])         # M::Vector{Float64}
             if 0 < slopes[i][j] < 1              # if in the triangle region of relaxed ReLU
                 #posind = M .> 0
-                neg .+= ℓ[j] * min.(M, 0) #-M .* !posind  # multiply by boolean to set the undesired values to 0.0
-                pos .+= ℓ[j] * max.(M, 0) #M .* posind
+                neg .+= ℓ[j] * min.(V, 0) #-M .* !posind  # multiply by boolean to set the undesired values to 0.0
+                pos .+= ℓ[j] * max.(V, 0) #M .* posind
             end
         end
     end
@@ -152,7 +149,6 @@ function input_layer_bounds(input_layer, input, ϵ)
     u = out1 + Δ
     return l, u
 end
-
 
 function new_μ(n_input, n_output, input_ReLU, WD)
     sub_μ = Vector{Vector{Float64}}(undef, n_input)

